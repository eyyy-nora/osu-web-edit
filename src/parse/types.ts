export interface ParsedBeatmap {
  General: {
    AudioFilename: string;
    AudioLeadIn: number;
    PreviewTime: number;
    CountDown: number;
    SampleSet: string;
    StackLaniency: number;
    Mode: number;
    LetterboxInBreaks: number;
    EpilepsyWarning: number;
    WidescreenStoryboard: number;
  };

  Editor: {
    Bookmarks: number[];
    DistanceSpacing: number;
    BeatDivisor: number;
    GridSize: number;
    TimelineZoom: number;
  };

  Metadata: {
    Title: string;
    TitleUnicode: string;
    Artist: string;
    ArtistUnicode: string;
    Creator: string;
    Version: string;
    Source: string;
    // Tags could be an string array
    Tags: string;
    BeatmapID: number;
    BeatmapSetID: number;
  };

  Difficulty: {
    ApproachRate: number;
    CircleSize: number;
    HPDrainRate: number;
    OverallDifficulty: number;
    SliderMultiplier: number;
    SliderTickRate: number;
  };

  Events: ParsedEvent[];
  TimingPoints: ParsedTimingPoint[];
  Colours: ParsedOsuColors;
  HitObjects: ParsedHitObject[];

}

export interface ParsedOsuColors {
  [index: `Combo${number}`]: [number, number, number];
  BackgroundColor: [number, number, number];
}

export interface ParsedHitSamples {
  normalSet: number;
  additionalSet: number;
  index: number;
  volume: number;
  filename?: string;
}

export interface ParsedPoint {
  x: number;
  y: number;
  anchor: boolean;
}

export interface HitObjectBase {
  x: number;
  y: number;
  time: number;
  hitSound: number;
  hitSample: ParsedHitSamples;
  skipColors: number;
  newCombo: boolean;
}

export interface ParsedHitCircle extends HitObjectBase {
  type: "circle";
}

export interface ParsedSliderBase extends HitObjectBase {
  type: "slider";
  slides: number;
  length: number;
  edgeSounds: number[];
  edgeSets: string[];
}

export interface ParsedLinearSlider extends ParsedSliderBase {
  sliderType: "linear";
  pathSegments: ParsedPoint[];
}

export interface ParsedPerfectCircleSlider extends ParsedSliderBase {
  sliderType: "circle";
  pathSegments: [ParsedPoint, ParsedPoint];
}

export interface ParsedBezierSlider extends ParsedSliderBase {
  sliderType: "bezier";
  pathSegments: ParsedPoint[];
}

export interface ParsedCatmullSlider extends ParsedSliderBase {
  sliderType: "catmull";
  pathSegments: ParsedPoint[];
}

export type ParsedSlider = ParsedLinearSlider | ParsedPerfectCircleSlider | ParsedBezierSlider | ParsedCatmullSlider;

export interface ParsedSpinner extends HitObjectBase {
  type: "spinner";
  end: number;
  x: 256;
  y: 192;
}

export interface ParsedHold extends HitObjectBase {
  type: "hold";
  end: number;
  y: 192;
}

export type ParsedHitObject = ParsedHitCircle | ParsedSlider | ParsedSpinner | ParsedHold;

export interface ParsedBackgroundEvent {
  type: "background";
  time: number;
  filename: string;
  x: number;
  y: number;
}

export interface ParsedVideoEvent {
  type: "video";
  time: number;
  filename: string;
  x: number;
  y: number;
}

export interface  ParsedBreakEvent {
  type: "break";
  time: number;
  end: number;
}

export type ParsedEvent = ParsedBackgroundEvent | ParsedVideoEvent | ParsedBreakEvent;

export interface ParsedTimingPoint {
  time: number;
  beatLength: number;
  meter: number;
  sampleSet: number;
  sampleIndex: number;
  volume: number;
  inherited: boolean;
  kiai: boolean;
  omitFirstBarLine: boolean;
}

export interface ParsedMapSet {
  difficulties: ParsedBeatmap[];
  files: Record<string, Blob>;
}
