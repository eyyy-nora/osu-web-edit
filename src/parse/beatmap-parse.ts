

export interface HitObject {
  x: number;
  y: number;
  SliderAnchors?: string[];
  Time: number;
  Type: number;
  HitSound: number;
  ObjectParameters: number;
  HitSample: string;
}
