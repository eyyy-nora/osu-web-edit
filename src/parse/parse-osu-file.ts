// noinspection JSBitwiseOperatorUsage

import axios from "axios";
import jszip from "jszip";
import { MapsetWithFolder } from "../context/mapset-context";
import {
  BIT0,
  BIT1,
  BIT2,
  BIT3,
  BIT4,
  BIT5,
  BIT6,
  BIT7
} from "../util/numbers";
import type {
  ParsedBeatmap,
  ParsedEvent,
  ParsedHitCircle,
  ParsedHitObject,
  ParsedHitSamples,
  ParsedHold,
  ParsedMapSet,
  ParsedPoint,
  ParsedSlider,
  ParsedSpinner,
  ParsedTimingPoint
} from "./types";

export function readMapset(file: File): Promise<MapsetWithFolder>  {
  return new Promise<MapsetWithFolder>((resolve, reject) => {
    const reader = new FileReader();
    reader.addEventListener("load", (ev) => {
      parseMapset(reader.result as ArrayBuffer).then(resolve).catch(reject);
    });
    reader.addEventListener("error", reject);
    reader.addEventListener("abort", reject);
    reader.readAsArrayBuffer(file);
  });
}

export async function downloadMapset(url: string): Promise<MapsetWithFolder>  {
  const { data } = await axios.get(url, { responseType: "blob" });
  return parseMapset(data);
}

export function mimeTypeFor(fileName: string): string {
  const ext = fileName.slice(fileName.lastIndexOf(".")).toLowerCase();
  switch (ext) {
    case ".mp3": return "audio/mp3";
    case ".wav": return "audio/wav";
    case ".ogg": return "audio/ogg";
    default: return "application/octet-stream";
  }
}


export async function parseMapset(file: Blob | Uint8Array | ArrayBuffer): Promise<MapsetWithFolder> {
  const zip = await jszip.loadAsync(file);

  const mapset: ParsedMapSet = { difficulties: [], files: {} };

  for (const [fileName, entry] of Object.entries(zip.files)) {
    if (/\.osu$/.test(fileName)) {
      const content = await entry.async("string");
      mapset.difficulties.push(parseOsuFile(content));
    } else {
      const type = mimeTypeFor(fileName)
      mapset.files[fileName] = new Blob([await entry.async("blob")], { type });
    }
  }

  return { mapSet: mapset, mapFolder: zip } as MapsetWithFolder;
}



const arraySections = ["HitObjects", "TimingPoints", "Events"];
export function parseOsuFile(data: string): ParsedBeatmap {
  const value = {};
  const lines = data.split(/[\r\n]+/);
  let section = null;

  for (const line of lines) {
    if (!line || line.startsWith(";") || line.startsWith("//")) continue; // comments

    if (line.startsWith("[")) {
      section = line.replace(/[\[\]]/g, "");
    } else if (section && arraySections.includes(section)) {
      value[section] ??= [];
      switch (section) {
        case "HitObjects": value[section]!.push(parseHitObject(line)); break;
        case "TimingPoints": value[section]!.push(parseTimingPoint(line)); break;
        case "Events": value[section]!.push(parseEvent(line)); break;
      }
    } else if (section && section === "Colours") {
      value[section] ??= {};
      const [key, val] = line.split(":", 2).map((str) => str.trim());
      const [r, g, b] = val.split(",").map(Number);
      value[section][key] = [r, g, b];
    } else if (line.includes(":") && !arraySections.includes(section)) {
      const [key, val] = line.split(":", 2).map((str) => str.trim());
      let parsedValue = val;
      try { parsedValue = JSON.parse(val); }
      catch {}
      if (section) (value[section] ??= {})[key] = parsedValue;
      else value[key] = parsedValue;
    }
  }

  return value as ParsedBeatmap;
}


export function parseTimingPoint(line: string): ParsedTimingPoint {
  const [time, beatLength, meter, sampleSet, sampleIndex, volume, uninerited, effects] = line.split(",").map(Number);
  const inherited = !uninerited;
  const kiai = !!(effects & BIT0);
  const omitFirstBarLine = !!(effects & BIT3);
  return { time, beatLength, meter, inherited, kiai, omitFirstBarLine, sampleIndex, sampleSet, volume };
}

export function parseEvent(line: string): ParsedEvent {
  const [type_, startTime, ...params] = line.split(",");
  switch (type_) {
    case "0":
      return { type: "background", time: +startTime, filename: params[0], x: +params[1], y: +params[2] };
    case "Video":
    case "1":
      return { type: "video", time: +startTime, filename: params[0], x: +params[1], y: +params[2] };
    case "2":
    case "Break":
      return { type: "break", time: +startTime, end: +params[0] };
  }
}


const firstFiveAsNumber = (part: string, index: number) => index <= 4 ? +part : part;
const firstFourAsNumber = (part: string, index: number) => index <= 3 ? +part : part;

export function parseHitObject(line: string): ParsedHitObject {
  const [x, y, time, type_, hitSound, ...rest] = line.split(",").map(firstFiveAsNumber) as [number, number, number, number, number, ...string[]];
  const newCombo = !!(type_ & BIT2);
  const skipColors = (type_ & (BIT4 | BIT5 | BIT6)) >> 4;
  const hitSample = parseHitSamples(rest[rest.length - 1]);

  let value: ParsedHitObject;

  if (type_ & BIT0) value = parseCircle();
  else if (type_ & BIT1) value = parseSlider(...rest as [string, string, string, string, string]);
  else if (type_ & BIT3) value = parseSpinner(rest[0]);
  else if (type_ & BIT7) value = parseHold(rest[0]);
  else throw new Error("Invalid object type " + type_);

  Object.assign(value, { x, y, time, newCombo, skipColors, hitSound, hitSample });

  return value;
}

export function parseHitSamples(sample: string): ParsedHitSamples {
  const [normalSet, additionalSet, index, volume, filename] = sample.split(":").map(firstFourAsNumber) as [number, number, number, number, string];
  return { normalSet, additionalSet, index, volume, filename };
}

export function parseSpinner(end: string): ParsedSpinner {
  return { type: "spinner", end: +end } as ParsedSpinner;
}

export function parseHold(end: string): ParsedHold {
  return { type: "hold", end: +end } as ParsedHold;
}

export function parseCircle(): ParsedHitCircle {
  return { type: "circle" } as ParsedHitCircle;
}

const sliderTypes = {
  "L": "linear",
  "P": "circle",
  "B": "bezier",
  "C": "catmull",
} as const;

export function parseSlider(curveTypeAndPoints: string, slides: string, length: string, edgeSounds: string, edgeSets: string): ParsedSlider {
  const [prefix, ...points] = curveTypeAndPoints.split("|");
  const sliderType = sliderTypes[prefix];
  const pathSegments = parseSliderPoints(points);
  return {
    type: "slider",
    sliderType,
    pathSegments,
    slides: +slides,
    length: +length,
    edgeSounds: edgeSounds?.split("|").map(sound => +sound) ?? [],
    edgeSets: edgeSets?.split("|") ?? [],
  } as ParsedSlider;
}

export function parseSliderPoints(points: string[]): ParsedPoint[] {
  const pathSegments: ParsedPoint[] = [];
  let last: [number, number] | undefined = undefined;
  for (const [x, y] of points.map(point => point.split(":").map(value => +value)) as [number, number][]) {
    if (last && last[0] === x && last[1] === y) pathSegments[pathSegments.length - 1].anchor = true;
    else {
      pathSegments.push({ x, y, anchor: false });
      last = [x, y];
    }
  }
  return pathSegments;
}

