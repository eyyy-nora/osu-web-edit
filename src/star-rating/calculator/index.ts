import BeatmapParser from "./BeatmapParser";
import DifficultyHitObjectCreator from "./DifficultyHitObjectCreator";
import StarRatingCalculator, { SRCalculatorResponse } from "./StarRatingCalculator";

const beatmapParser = new BeatmapParser();
const difficultyHitObjectCreator = new DifficultyHitObjectCreator();
const starRatingCalculator = new StarRatingCalculator();
let Beatmap = null;

export function calculateStarRating(dotOsuFile: string, mods?: string[], allCombinations?: boolean, returnAllDifficultyValues: boolean = false): object {
    if (dotOsuFile === null)
        throw new Error("Difficulty Calculation Failed: Provided file is empty");

    mods = parseMods(mods);
    let output = { };
    if (!allCombinations) {
        const label = mods.length > 0 ? mods.join('') : "nomod";
        const response = calculateNextModCombination(dotOsuFile, mods, true);
        output[label] = returnAllDifficultyValues ? response : response.total;
        return output;
    }
    else {
        const allModCombinations = getAllModCombinations();
        allModCombinations.forEach(combi => {
            const label = combi.mods.length > 0 ? combi.mods.join('') : "nomod";
            const response = calculateNextModCombination(dotOsuFile, combi.mods, combi.reParse);
            output[label] = returnAllDifficultyValues ? response : response.total;
        });
        return output;
    }
}

function calculateNextModCombination(dotOsuFile: string, mods: string[], reParse: boolean): SRCalculatorResponse {
    if (reParse)
        Beatmap = beatmapParser.parseBeatmap(dotOsuFile, mods);

    const timeRate = getTimeRate(mods);
    const difficultyHitObjects = difficultyHitObjectCreator.convertToDifficultyHitObjects(Beatmap.HitObjects, timeRate);
    return starRatingCalculator.calculate(difficultyHitObjects, timeRate);
};

function parseMods(mods: string[]): string[] {
    if (mods === undefined)
        return [];
    return mods;
};

function getTimeRate(mods: string[]): number {
    if (mods.includes("DT"))
        return 1.5;
    if (mods.includes("HT"))
        return 0.75;
    return 1;
};

function getAllModCombinations(): Array<{ mods: string[], reParse: boolean }> {
    return [
        { mods: [], reParse: true },
        { mods: ["DT"], reParse: false },
        { mods: ["HT"], reParse: false },
        { mods: ["HR"], reParse: true },
        { mods: ["HR", "DT"], reParse: false },
        { mods: ["HR", "HT"], reParse: false },
        { mods: ["EZ"], reParse: true },
        { mods: ["EZ", "DT"], reParse: false },
        { mods: ["EZ", "HT"], reParse: false }
    ];
};