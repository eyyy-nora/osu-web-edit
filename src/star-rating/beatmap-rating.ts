import { ParsedBeatmap, ParsedHitObject } from "../parse/types";
import { createAimAccumulator } from "./aim";
import { createSpeedAccumulator } from "./speed";
import { BeatmapMod, SkillAccumulatorMap, SkillResultMap, Skills } from "./types";


export function calculateBeatmapRating<Accumulators extends SkillAccumulatorMap>(
  beatmap: ParsedBeatmap,
  accumulators: Accumulators,
  mods: BeatmapMod[] = [],
): SkillResultMap<Skills<Accumulators>> {
  const sectionBaseLength = 400, difficultyMultiplier = 0.0675;
  const sectionLength = sectionBaseLength * (mods.includes("DT") ? 1.5 : mods.includes("HT") ? .75 : 1);

  if (!("aim" in accumulators)) accumulators.aim = createAimAccumulator(beatmap);
  if (!("speed" in accumulators)) accumulators.aim = createSpeedAccumulator(beatmap);

  for (const accumulator of Object.values(accumulators))
    accumulator.init?.(mods);

  for (const section of hitObjectSections(beatmap.HitObjects, sectionLength))
    for (const accumulator of Object.values(accumulators))
      accumulator.processSection(section);

  const results: SkillResultMap<Skills<Accumulators>> =
    Object.fromEntries(Object.entries(accumulators).map(([key, value]) => {
      return [key, Math.sqrt(value.accumulate()) * difficultyMultiplier];
    })) as never;

  results.stars = results.aim + results.speed + Math.abs(results.aim - results.speed) / 2;
  return results;
}

export function startTime(objects: ParsedHitObject[]): number {
  return objects?.[0].time ?? 0;
}

export function sectionBounds(time: number, sectionLength: number): [number, number] {
  const start = Math.floor(time / sectionLength);
  return [start * sectionLength, (start + 1) * sectionLength];
}

export function* hitObjectSections(objects: ParsedHitObject[], sectionLength: number) {
  let [, sectionEnd] = sectionBounds(startTime(objects), sectionLength);
  let sectionStartIndex = 0;

  for (let i = 0; i < objects.length; ++i)
    if (objects[i].time > sectionEnd) {
      yield objects.slice(sectionStartIndex, i - 1);
      sectionStartIndex = i;
      [, sectionEnd] = sectionBounds(objects[i].time, sectionLength);
    }

  yield objects.slice(sectionStartIndex, objects.length);
}
