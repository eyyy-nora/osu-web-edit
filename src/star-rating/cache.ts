import { calculateStarRating } from "./calculator/index";
import { BeatmapObject } from "../../src/io/types/beatmap/objects/object";
import { Beatmap } from "../../src/io/types/beatmap/beatmap";
import { serializeBeatmap } from "../../src/io/serializer/beatmap";

let starRatingCache: BeatmapStarRating[] = [];

export async function cacheStarRating(beatmap: Beatmap): Promise<number> {
  const NUMBERS_AFTER_DOT = 2;


  const serializedMap = await serializeBeatmap(beatmap);

  const starRating = parseFloat(

    calculateStarRating(serializedMap)["nomod"].toFixed(NUMBERS_AFTER_DOT)

  );

  saveToCache(beatmap, starRating);

  return starRating;
}

function saveToCache(beatmap: Beatmap, starRating: number): void {
  let cacheEntry: BeatmapStarRating = {

    beatmapSetId: beatmap.metadata.beatmapSetID,
    version: beatmap.metadata.version,
    objects: beatmap.objects,
    starRating: starRating

  };

  starRatingCache.push(cacheEntry);
}

export interface BeatmapStarRating {
  beatmapSetId: number;
  version: string;
  objects: BeatmapObject[];
  starRating: number;
}

export function getCachedStars(): BeatmapStarRating[] {
  return starRatingCache;
}

export function flushCache(): void {
  starRatingCache = []
}
