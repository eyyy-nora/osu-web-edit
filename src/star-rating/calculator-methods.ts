import { calculateStarRating } from "./calculator/index";

export function computeStarRating(rawBeatmap: string) {
  return parseFloat(calculateStarRating(rawBeatmap)["nomod"].toFixed(2));
}
