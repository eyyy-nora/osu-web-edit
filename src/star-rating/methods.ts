import { ParsedBeatmap } from "../parse/types";
import { BeatmapSRCache, saveSRCache, getCachedSR } from "./cache";

// TODO: When the function to turn parsedBeatmaps to .osu is finished, refactor
// this file to use it instead of receiving rawBeatmap as a parameter.
export function computeStarRating(rawBeatmap: string, parsedBeatmap: ParsedBeatmap): number {
  let beatmapsCache = getCachedSR();

  if (beatmapsCache.length > 0) {
    for (let i = 0; i < beatmapsCache.length; i++) {
      let cachedBeatmap = beatmapsCache[i];

      if (beatmapsAreEqual(cachedBeatmap, parsedBeatmap)) 
        return cachedBeatmap.StarRating;

      else if (isLastLoopIteration(i, beatmapsCache.length)) 
        return saveSRCache(rawBeatmap, parsedBeatmap);

    }
  } else {
    return saveSRCache(rawBeatmap, parsedBeatmap);
  }

}


function isLastLoopIteration(iterator: number, length: number): boolean {
  return iterator === (length - 1)
}

function beatmapsAreEqual(beatmapInCache: BeatmapSRCache, checkedBeatmap: ParsedBeatmap ): boolean {
  return (
    beatmapInCache.BeatmapSetID === checkedBeatmap.Metadata.BeatmapSetID &&
    beatmapInCache.Version === checkedBeatmap.Metadata.Version &&
    beatmapInCache.HitObjects === checkedBeatmap.HitObjects
  );
} 

