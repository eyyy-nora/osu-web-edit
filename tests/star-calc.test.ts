import { parseOsuFile } from "../src/parse/parse-osu-file";
import fs from "fs";
import { computeStarRating } from "../src/star-rating/methods";
import { flushCache, getCachedSR } from "../src/star-rating/cache";

const dotOsuData = fs.readFileSync(__dirname + "/assets/testfile.osu");
const dotOsuDataAlt = fs.readFileSync(__dirname + "/assets/testfile-alt.osu");

const parsedDotOsu = parseOsuFile(dotOsuData.toString())
const parsedDotOsuAlt = parseOsuFile(dotOsuDataAlt.toString())


test('StarRating Calculation', () => {
  let beatmapStarRating = computeStarRating(dotOsuData.toString(), parsedDotOsu);

  expect(beatmapStarRating).toMatchSnapshot();

  flushCache();
});

test('StarRating Cache: Cache being written', () => {
  let beatmapStarRating = computeStarRating(dotOsuData.toString(), parsedDotOsu);

  let cachedBeatmaps = getCachedSR();

  expect(cachedBeatmaps.length).toBe(1);

  expect(cachedBeatmaps[0].StarRating).toBe(beatmapStarRating);
  expect(cachedBeatmaps[0].BeatmapSetID).toBe(parsedDotOsu.Metadata.BeatmapSetID);
  expect(cachedBeatmaps[0].Version).toBe(parsedDotOsu.Metadata.Version);

  flushCache();
});

test('StarRating Cache: Cache not being written when the same beatmap is computed twice', () => {
  computeStarRating(dotOsuData.toString(), parsedDotOsu);
  let beatmapStarRating = computeStarRating(dotOsuData.toString(), parsedDotOsu);

  let cachedBeatmaps = getCachedSR();

  expect(cachedBeatmaps.length).toBe(1);
  expect(cachedBeatmaps[0].StarRating).toBe(beatmapStarRating);

  flushCache();
});

test('StarRating Cache: Beatmap not being found in a cache already filled', () => {
  computeStarRating(dotOsuDataAlt.toString(), parsedDotOsuAlt);
  let beatmapStarRating = computeStarRating(dotOsuData.toString(), parsedDotOsu);

  let cachedBeatmaps = getCachedSR();

  expect(cachedBeatmaps.length).toBe(2);
  expect(cachedBeatmaps[1].StarRating).toBe(beatmapStarRating);

  flushCache();
})
