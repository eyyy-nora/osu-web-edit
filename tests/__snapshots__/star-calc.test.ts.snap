// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StarRating Calculation 1`] = `8.18`;
